miden-tx/src/prover/mod.rs

```diff
@@ -57,24 +64,44 @@ impl TransactionProver {
                 .map_err(TransactionProverError::ProveTransactionProgramFailed)?;
 
         // extract transaction outputs and process transaction data
-        let (advice_provider, _event_handler) = host.into_parts();
+        let (advice_provider, account_delta) = host.into_parts();
         let (_, map, _) = advice_provider.into_parts();
         let tx_outputs = TransactionKernel::parse_transaction_outputs(&stack_outputs, &map.into())
             .map_err(TransactionProverError::InvalidTransactionOutput)?;
 
-        Ok(ProvenTransaction::new(
-            account_id,
-            if tx_witness.account().is_new() {
-                Digest::default()
-            } else {
-                initial_account_hash
+        let initial_hash = if tx_witness.account().is_new() {
+            Digest::default()
+        } else {
+            initial_account_hash
+        };
+
+        let builder = ProvenTransactionBuilder::new()
+            .account_id(account_id)
+            .initial_account_hash(initial_hash)
+            .final_account_hash(tx_outputs.account.hash())
+            .add_input_notes(input_notes)
+            .add_output_notes(tx_outputs.output_notes.into_iter().map(NoteEnvelope::from))
+            .block_ref(block_hash)
+            .proof(proof);
+
+        let builder = match tx_script_root {
+            Some(tx_script_root) => builder.tx_script_root(tx_script_root),
+            _ => builder,
+        };
+
+        let builder = match on_chain_account {
+            true => {
+                let account_details = if tx_witness.account().is_new() {
+                    AccountDetails::Full(tx_witness.account().clone())
+                } else {
+                    AccountDetails::Delta(account_delta)
+                };
+
+                builder.account_details(account_details)
             },
-            tx_outputs.account.hash(),
-            input_notes,
-            tx_outputs.output_notes.into(),
-            tx_script_root,
-            block_hash,
-            proof,
-        ))
+            false => builder,
+        };
```

Comment:
I'm not particularly found of this API. But as the things are currently set, this seems to be the best place to have it.

An alternative would be to decouple the `ProvenTransaction` from the account/note state. But at the same time I don't think that would be worth the effort.